spec/integration/autobahn_intg_spec.rb

```diff
@@ -319,6 +319,20 @@ def create_transport_system(name, options={})
         await_delivery
         @queues.map { |q| q.status.first }.reduce(:+).should == 1
       end
+
+      it 'does not deliver any new messages after it has been unsubscribed' do
+        unsubscribed = Autobahn::Concurrency::Semaphore.new(0)
+        @consumer.subscribe do |headers, message|
+          begin
+            unsubscribed.acquire
+            headers.ack
+          rescue java.lang.InterruptedException
+          end
+        end
+        unsubscribed.release
+        @consumer.unsubscribe!
+        unsubscribed.has_queued_threads.should be_false
+      end
```

Comment:
I think the problem here is that you try to combine too many things into one test. Assuming you only wanted to assert what the test description says, you would essentially want to just increment a counter from subscribe, and then set the counter to zero right after the unsubscribe call. A test like that would obviously require that you concurrently enqueue a bunch of things, to give a time-window after the unsubscribe-call finished where new calls could take place. Such a test would obviously not be deterministic.

I think you could replace the test with some more specific tests:
 - waits for current delivery to finish. by making the subscribe body increment a counters on entry and exit, with a little sleep in the middle, you could assert that the two counters should be equal (and nonzero) afterwards.
 - interrupts a delivery if it takes too long. do a semaphore acquire like you do in this test, but don't release it.
 - verify that it terminates the subscription thread(s). could be implemented by collecting `Thread.current` from the delivery callback, and asserting that the resulting threads aren't alive after the unsubscribe call.

None of these tests really capture the full desired behavior, but for that you would need to implement unit tests where you could actually inject the thread pool and whatnot. That seems like a much bigger change, though.